gobblin-data-management/src/main/java/gobblin/data/management/copy/RecursiveCopyableDataset.java

```diff
@@ -51,29 +73,90 @@ public RecursiveCopyableDataset(final FileSystem fs, Path rootPath, Properties p
+    Map<Path, FileStatus> filesInSource = createPathMap(getFilesAtPath(this.fs, this.rootPath, this.pathFilter), this.rootPath);
+    Map<Path, FileStatus> filesInTarget = createPathMap(getFilesAtPath(targetFs, targetPath, this.pathFilter), targetPath);
 
-    List<CopyableFile> copyableFiles = Lists.newArrayList();
+    Set<Path> inSourceNotInTarget = Sets.difference(filesInSource.keySet(), filesInTarget.keySet());
+    Set<Path> inTargetNotInSource = Sets.difference(filesInTarget.keySet(), filesInSource.keySet());
+    Set<Path> inBoth = Sets.intersection(filesInSource.keySet(), filesInTarget.keySet());
+
+    List<CopyEntity> copyEntities = Lists.newArrayList();
 
-    for (FileStatus file : files) {
+    List<Path> toCopy = Lists.newArrayList();
+    List<Path> toDelete = Lists.newArrayList();
+
+    for (Path path : inBoth) {
+      if (!sameFile(filesInSource.get(path), filesInTarget.get(path))) {
+        toCopy.add(path);
+        toDelete.add(path);
+      }
+    }
+    if (!this.update && !toCopy.isEmpty()) {
+      throw new IOException("Some files need to be copied but they already exist in the destination. "
+          + "Aborting because not running in update mode.");
+    }
+
+    for (Path path : inSourceNotInTarget) {
+      toCopy.add(path);
+    }
+
+    if (this.delete) {
+      for (Path path : inTargetNotInSource) {
+        toDelete.add(path);
+      }
+    }
+
+    List<CopyableFile> copyableFiles = Lists.newArrayList();
+    for (Path path : toCopy) {
+      FileStatus file = filesInSource.get(path);
       Path filePathRelativeToSearchPath = PathUtils.relativizePath(file.getPath(), nonGlobSearchPath);
-      Path targetPath = new Path(configuration.getPublishDir(), filePathRelativeToSearchPath);
+      Path thisTargetPath = new Path(configuration.getPublishDir(), filePathRelativeToSearchPath);
 
-      copyableFiles.add(CopyableFile.fromOriginAndDestination(this.fs, file, targetPath, configuration)
-          .fileSet(file.getPath().getParent().toString())
+      copyableFiles.add(CopyableFile.fromOriginAndDestination(this.fs, file, thisTargetPath, configuration)
+          .fileSet(datasetURN())
           .ancestorsOwnerAndPermission(CopyableFile.resolveReplicatedOwnerAndPermissionsRecursively(this.fs,
               file.getPath().getParent(), nonGlobSearchPath, configuration))
           .build());
     }
-    return this.copyableFileFilter.filter(this.fs, targetFs, copyableFiles);
+    copyEntities.addAll(this.copyableFileFilter.filter(this.fs, targetFs, copyableFiles));
+
+    if (!toDelete.isEmpty()) {
+      List<FileStatus> statusesToDelete = Lists.newArrayList();
+      for (Path path : toDelete) {
+        statusesToDelete.add(filesInTarget.get(path));
+      }
+
+      CommitStep step =
+          new DeleteFileCommitStep(targetFs, statusesToDelete, this.properties,
+              this.deleteEmptyDirectories ? Optional.of(targetPath) : Optional.<Path>absent());
```

Comment:
how is deleteEmptyDirectories been used? I thought somewhere, we need to check that FileStatus is empty dir, right?